astacus/coordinator/plugins/cassandra/restore_steps.py

```diff
@@ -77,6 +87,66 @@ def find_matching_cassandra_index(self, backup_node: CassandraManifestNode) -> O
         return None
 
 
+@dataclass
+class UploadFinalDeltaStep(Step[List[ipc.BackupManifest]]):
+    json_storage: AsyncJsonStorage
+    storage_name: str
+
+    async def run_step(self, cluster: Cluster, context: StepsContext) -> List[ipc.BackupManifest]:
+        replaced_nodes = context.get_result(StopReplacedNodesStep)
+        assert len(replaced_nodes) == 1, "Multiple node replacement not supported"
+        replaced_node = replaced_nodes[0]
+
+        snapshot_results = await self.snapshot_delta(replaced_node, cluster=cluster, context=context)
+        assert len(snapshot_results) == 1
+        assert snapshot_results[0].hashes is not None
+        upload_results = await self.upload_delta(replaced_node, snapshot_results[0].hashes, cluster=cluster)
+        return await self.upload_manifest(snapshot_results, upload_results, context=context)
+
+    async def snapshot_delta(
+        self, node: CoordinatorNode, *, cluster: Cluster, context: StepsContext
+    ) -> List[ipc.SnapshotResult]:
+        snapshot_results = await SnapshotStep(
+            snapshot_groups=delta_snapshot_groups(), snapshot_request="delta/snapshot", nodes_to_snapshot=[node]
+        ).run_step(cluster, context)
```

Comment:
I don't think step should call other steps.
Steps are not meant to be used as arbitrary function calls, they have impact on what is logged and how Astacus reports its progress for instance.
If you need to share code between steps, extract a function and use that function in the two steps.

Reply:
IIRC the logging and progress are provided by whatever calls the step, but I agree, calling the step directly looks like a WTF. Extracted the necessary common logic.